### PR TITLE
Added reach estimator based on FirstMomementEstimator that add global noise

### PR DESCRIPTION
**Pull request by kungfucraig**
Extended FirstMomentEstimator, such that when a noiser is provider, noise is added to the number of active buckets before performing the cardinality estimation. This support the global DP scenario implemented by our MPC protocol.

### CODE REVIEW

File: src/estimators/bloom_filters.py [hunk 1]
```diff
@@ -419,6 +421,53 @@ def __call__(self, sketch_list):
...
+    union = sketch_list[0]
+    for cur_sketch in sketch_list[1:]:
+      union.assert_compatible(cur_sketch)
+      union.sketch = 1 - (1 - union.sketch) * (1 - cur_sketch.sketch)
+
+    noiser = GeometricEstimateNoiser(
+        self.epsilon, random_state=self.random_state)
+    noise = noiser(0.)
+
+    a = union.decay_rate
```
Comment (kungfucraig) [inline src/estimators/bloom_filters.py:450]: This we agree to this approach I'll do some refactoring to pull out the code that this shares with the FirstMomentEstimator, or make some of the FirstMomentEstimator's methods public.

File: src/estimators/bloom_filters.py [hunk 2]
```diff
@@ -419,6 +421,53 @@ def __call__(self, sketch_list):
         union, self._weights)]
 
 
+class FirstMomentGlobalNoiseEstimator(EstimatorBase):
+  """This estimator unions the input sketches, generate geometric noise,
+     appends that many fake buckets to the sketch union and calls the
+     FirstMomentEstimator.
```
Comment (shengma-1) [inline src/estimators/bloom_filters.py:427]: This only supports exp-adbf, right? Do you want to make this more general for all adbf families?

An alternative is to extend FirstMomentNoiseEstimator to take a global_estimator_noiser, which, if defined, will be added to #ones inside of FirstMomentNoiseEstimator.

Comment (kungfucraig) [reply src/estimators/bloom_filters.py:427]: Yeah, I think that's a reasonable way to refactor. Will send an update.

Comment (huangxichen1) [reply src/estimators/bloom_filters.py:427]: Another approach is to change the AnyDistributionBloomFilter, i.e., add an optional global noiser argument to the constructor, and create a property method for the number of one bits (the `x` in the exp and log bloom filter estimator, and the `n` in the geo bloom filter). In this way, we don't need to modify all the estimator methods. Please let me know if this makes sense.

Thanks Craig!

Comment (kungfucraig) [reply src/estimators/bloom_filters.py]: PTAL. Refactored FirstMomentEstimator based on conversation yesterday.

File: src/estimators/bloom_filters.py [hunk 3]
```diff
@@ -290,12 +292,13 @@ class FirstMomentEstimator(EstimatorBase):
   METHOD_EXP = "exp"
   METHOD_ANY = "any"
 
-  def __init__(self, method, denoiser=None, weights=None):
+  def __init__(self, method, denoiser=None, noiser=None, weights=None):
```
Comment (matthewclegg) [inline src/estimators/bloom_filters.py:295]: I suppose if you set the default value for noiser=lambda x:x, you could save a few lines of code.

Comment (kungfucraig) [reply src/estimators/bloom_filters.py:295]: Setting the default value of the argument makes my error check unworkable, however I did use the lambda trick in the body of __init__, which did save a few lines of repeated code in other functions. Thanks!

**Comment by kungfucraig** [comment]
@AaronRanAn @shengma-1 

I need to get this submitted today, so given other reviews I'm going to submit now. Please feel to review later if you'd like.

**Pull request merged**